src/synapse-plugins/converter-plugin/converter-plugin-backend.vala

```diff
@@ -0,0 +1,372 @@
+                                results += result;
+                            }
+                        }
+                    }
+                }
+            }
+
+            return results;
+        }
+
+        private ResultData? calculate_conversion_data (
+            double num,
+            UnitMatch match1,
+            UnitMatch match2) {
+
+            Unit u1 = match1.unit, u2 = match2.unit;
+            int dim1 = match1.dimension, dim2 = match2.dimension;
+            double factor1 = match1.factor (), factor2 = match2.factor (); // Takes into account dimension
+            string descr1 = match1.description (), descr2 = match2.description ();
+            bool same_system = u1.system == u2.system;
+
+            // Find factor for each unit to a common base unit (taking into account dimensionality and prefixes)
+            // If both given units are in the same system stop at the base unit for that system, otherwise
+            // convert to SI.
+            Unit? parent = u1; // Parent should only be null in the case of an error in the data structure.
+            int parent_dimension = 1;
+            debug ("finding root of %s - start dimension %i, start factor %f", u1.uid, dim1, factor1);
+            while (parent != null &&
+                   parent.base_unit != "" &&
+                   (!same_system || parent.system == u1.system)) {
+
+                parent = find_parent_unit (parent.base_unit, out parent_dimension);
+                var pfactor = parent.get_factor ();
+                debug ("parent1 %s parent_dimension1 %i, parent_factor1 %f",
+                       parent.uid, parent_dimension, pfactor
+                );
+
+                dim1 *= parent_dimension;
+                debug ("Dim1 now %i", dim1);
+                for (int i = 0; i < dim1; i++) {
+                    factor1 *= pfactor;
+                    debug ("Factor1 now %f", factor1);
+                }
+            }
+
+            if (parent == null) {
+                return null;
+            }
+
+            parent_dimension = 1;
+            var ultimate_parent1 = parent.uid;
+            parent = u2;
+            debug ("finding root of %s - start dimension %i, start factor2 %f", u2.uid, dim2, factor2);
+            while (parent != null &&
+                   parent.base_unit != "" &&
+                   (!same_system || parent.system == u2.system)) {
+
+                parent = find_parent_unit (parent.base_unit, out parent_dimension);
+                var pfactor = parent.get_factor ();
+                debug ("parent2 %s parent_dimension2 %i, parent_factor2 %f",
+                    parent.uid, parent_dimension, pfactor
+                );
+                dim2 *= parent_dimension;
+                debug ("Dim2 now %i", dim2);
+                for (int i = 0; i < dim2; i++) {
+                    factor2 *= pfactor;
+                    debug ("Factor2 now %f", factor2);
+                }
+            }
```

Comment:
This is very similar to the parent1 logic, maybe it would be better to move it to a function which you could reuse twice?